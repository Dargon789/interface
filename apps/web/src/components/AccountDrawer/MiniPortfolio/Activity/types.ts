import { Currency } from '@uniswap/sdk-core'
import { GraphQLApi } from '@universe/api'
import { UniverseChainId } from 'uniswap/src/features/chains/types'
import {
  TransactionOptions,
  TransactionStatus,
  UniswapXOrderDetails,
} from 'uniswap/src/features/transactions/types/transactionDetails'

/**
 * TODO: refactor parsing / Activity so that all Activity Types can have a detail sheet.
 */

export type Activity = {
  id: string
  hash?: string
  chainId: UniverseChainId
  outputChainId?: UniverseChainId
  status: TransactionStatus
  offchainOrderDetails?: UniswapXOrderDetails
  statusMessage?: string
  timestamp: number
  title: string
  descriptor?: string | JSX.Element
  logos?: Array<string | undefined>
  // TODO(WEB-3839): replace Currency with CurrencyInfo
  currencies?: Array<Currency | undefined>
  otherAccount?: string
  from: string
  options?: TransactionOptions
  prefixIconSrc?: string
  suffixIconSrc?: string
  isSpam?: boolean
  type?: GraphQLApi.TransactionType
}

export type ActivityMap = { [id: string]: Activity | undefined }
