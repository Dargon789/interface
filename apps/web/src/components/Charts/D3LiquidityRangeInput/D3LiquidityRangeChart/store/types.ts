import { GraphQLApi } from '@universe/api'
import { TickAlignment } from 'components/Charts/D3LiquidityRangeInput/D3LiquidityRangeChart/utils/priceToY'
import { ChartEntry } from 'components/Charts/LiquidityRangeInput/types'
import { PriceChartData } from 'components/Charts/PriceChart'
import * as d3 from 'd3'
import { UseSporeColorsReturn } from 'ui/src/hooks/useSporeColors'

export type ChartState = {
  dimensions: {
    width: number
    height: number
  }
  dragCurrentTick?: ChartEntry
  dragCurrentY?: number
  dragStartTick?: ChartEntry
  dragStartY: number | null
  dynamicZoomMin: number
  hoveredTick?: ChartEntry
  hoveredY?: number
  initialViewSet: boolean
  isChartHovered?: boolean
  isFullRange: boolean
  maxPrice?: number
  minPrice?: number
  panY: number
  selectedHistoryDuration: GraphQLApi.HistoryDuration
  selectedPriceStrategy?: DefaultPriceStrategy
  zoomLevel: number
}

export type AnimationParams = {
  targetZoom: number
  targetPan: number
  targetMinPrice?: number
  targetMaxPrice?: number
  duration?: number
}

export type RenderingContext = {
  colors: UseSporeColorsReturn
  dimensions: {
    width: number
    height: number
  }
  priceData: PriceChartData[]
  liquidityData: ChartEntry[]
  tickScale: ((tick: string) => number) & {
    domain: () => string[]
    bandwidth: () => number
    range: () => [number, number]
  }
  priceToY: ({ price, tickAlignment }: { price: number; tickAlignment?: TickAlignment }) => number
  yToPrice: (y: number) => number
}

export enum DefaultPriceStrategy {
  STABLE = 'stable',
  WIDE = 'wide',
  ONE_SIDED_UPPER = 'one_sided_upper',
  ONE_SIDED_LOWER = 'one_sided_lower',
}

export interface Renderer {
  draw(): void
}

type Renderers = {
  priceLineRenderer: Renderer | null
  liquidityBarsRenderer: Renderer | null
  liquidityRangeAreaRenderer: Renderer | null
  minMaxPriceLineRenderer: Renderer | null
  scrollbarContainerRenderer: Renderer | null
  minMaxPriceIndicatorsRenderer: Renderer | null
  currentTickRenderer: Renderer | null
  liquidityBarsOverlayRenderer: Renderer | null
  timescaleRenderer: Renderer | null
}

export type ChartActions = {
  setChartState: (state: Partial<ChartState>) => void
  setPriceStrategy: (strategy: DefaultPriceStrategy) => void
  setTimePeriod: (timePeriod: GraphQLApi.HistoryDuration) => void
  updateDimensions: (dimensions: { width: number; height: number }) => void
  handlePriceChange: (changeType: 'min' | 'max', price?: number) => void
  initializeView: (params?: { minPrice: number | null; maxPrice: number | null }) => void
  initializeRenderers: ({
    g,
    timescaleG,
    context,
  }: {
    g: d3.Selection<SVGGElement, unknown, null, undefined>
    timescaleG: d3.Selection<SVGGElement, unknown, null, undefined>
    context: RenderingContext
  }) => void
  createHandleDragBehavior: (lineType: 'min' | 'max') => d3.DragBehavior<any, unknown, unknown>
  createTickBasedDragBehavior: () => d3.DragBehavior<any, unknown, unknown>
  centerRange: () => void
  zoom: (targetZoom: number) => void
  zoomIn: () => void
  zoomOut: () => void
  reset: (params?: { animate?: boolean; minPrice?: number; maxPrice?: number }) => void
  drawAll: () => void
  animateToState: (params: AnimationParams) => void
  incrementMax: () => void
  decrementMax: () => void
  incrementMin: () => void
  decrementMin: () => void
}

export type ChartStoreState = ChartState & {
  renderers: Renderers
  renderingContext: RenderingContext | null
  dynamicZoomMin: number
  actions: ChartActions
}
