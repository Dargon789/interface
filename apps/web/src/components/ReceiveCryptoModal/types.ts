export enum ReceiveModalState {
  DEFAULT = 0,
  CEX_TRANSFER = 1,
  QR_CODE = 2,
}

export type ReceiveCryptoModalInitialState =
  | { modalState: Exclude<ReceiveModalState, ReceiveModalState.QR_CODE> | undefined; qrCodeAddress?: undefined }
  | { modalState: ReceiveModalState.QR_CODE; qrCodeAddress: string }
