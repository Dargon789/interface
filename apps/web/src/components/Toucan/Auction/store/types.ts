import { UniverseChainId } from 'uniswap/src/features/chains/types'

// potentially missing clearing price
export interface AuctionDetails {
  auctionId: string
  chainId: UniverseChainId
  tokenSymbol: string
  tokenAddress: string
  tokenName: string // this is missing from what Rob sent
  logoUrl: string // this is missing from what Rob sent
  creatorAddress: string
  startBlock: number
  endBlock: number
  totalSupply: string
  tickSize: string
  graduationThreshold: number
  bidTokenAddress: string
}

export interface CheckpointData {
  clearingPrice: string
  cumulativeMps: number
}

export enum DisplayMode {
  VALUATION = 'VALUATION',
  TOKEN_PRICE = 'TOKEN_PRICE',
}

interface AuctionState {
  auctionDetails: AuctionDetails | null
  checkpointData: CheckpointData | null
  tokenColor?: string
  displayMode: DisplayMode
}

interface AuctionActions {
  setTokenColor: (color?: string) => void
  setDisplayMode: (mode: DisplayMode) => void
}

export type AuctionStoreState = AuctionState & {
  actions: AuctionActions
}
