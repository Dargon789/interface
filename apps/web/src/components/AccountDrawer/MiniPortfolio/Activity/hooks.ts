import { useAssetActivity } from 'appGraphql/data/apollo/AssetActivityProvider'
import { TradingApi } from '@universe/api'
import { useLocalActivities } from 'components/AccountDrawer/MiniPortfolio/Activity/parseLocal'
import { parseRemoteActivities } from 'components/AccountDrawer/MiniPortfolio/Activity/parseRemote'
import { Activity, ActivityMap } from 'components/AccountDrawer/MiniPortfolio/Activity/types'
import { getActivityNonce, hasEncodedOrder, haveSameNonce } from 'components/AccountDrawer/MiniPortfolio/Activity/utils'
import { useCreateCancelTransactionRequest } from 'components/AccountDrawer/MiniPortfolio/Activity/utils/cancel'
import { GasFeeResult, GasSpeed, useTransactionGasFee } from 'hooks/useTransactionGasFee'
import { useEffect, useMemo } from 'react'
import { usePendingTransactions, usePendingUniswapXOrders, useTransactionCanceller } from 'state/transactions/hooks'
import { UniverseChainId } from 'uniswap/src/features/chains/types'
import { useLocalizationContext } from 'uniswap/src/features/language/LocalizationContext'
import { isUniswapX } from 'uniswap/src/features/transactions/swap/utils/routing'
import { TransactionStatus, UniswapXOrderDetails } from 'uniswap/src/features/transactions/types/transactionDetails'

/** Detects transactions from same account with the same nonce and different hash */
function findCancelTx({
  localActivity,
  remoteMap,
  account,
}: {
  localActivity: Activity
  remoteMap: ActivityMap
  account: string
}): string | undefined {
  // Skip activities without nonce or non-pending status
  if (!getActivityNonce(localActivity) || localActivity.status !== TransactionStatus.Pending) {
    return undefined
  }

  for (const remoteTx of Object.values(remoteMap)) {
    if (!remoteTx) {
      continue
    }

    // A pending tx is 'cancelled' when another tx with the same account & nonce but different hash makes it on chain
    if (
      haveSameNonce(localActivity, remoteTx) &&
      remoteTx.from.toLowerCase() === account.toLowerCase() &&
      remoteTx.hash &&
      localActivity.hash &&
      remoteTx.hash.toLowerCase() !== localActivity.hash.toLowerCase() &&
      remoteTx.chainId === localActivity.chainId
    ) {
      return remoteTx.hash
    }
  }

  return undefined
}

/** Deduplicates local and remote activities */
export function combineActivities(localMap: ActivityMap = {}, remoteMap: ActivityMap = {}): Array<Activity> {
  const activityIds = [...new Set([...Object.keys(localMap), ...Object.keys(remoteMap)])]

  return activityIds.reduce((acc: Array<Activity>, id) => {
    const localActivity = localMap[id]
    const remoteActivity = remoteMap[id]

    // Skip if both activities are undefined
    if (!localActivity && !remoteActivity) {
      return acc
    }

    // If only one exists, use that one
    if (!localActivity) {
      acc.push(remoteActivity as Activity)
      return acc
    }
    if (!remoteActivity) {
      acc.push(localActivity)
      return acc
    }

    // Both exist - apply merging logic
    if (localActivity.status === TransactionStatus.Canceled) {
      // Hides misleading activities caused by cross-chain nonce collisions previously being incorrectly labelled as cancelled txs in redux
      // If there is no remote activity fallback to local activity
      if (localActivity.chainId !== remoteActivity.chainId) {
        acc.push(remoteActivity)
        return acc
      }
      // Remote data only contains data of the cancel tx, rather than the original tx, so we prefer local data here
      acc.push(localActivity)
    } else {
      // Generally prefer remote values to local value because i.e. remote swap amounts are on-chain rather than client-estimated
      acc.push({ ...localActivity, ...remoteActivity } as Activity)
    }

    return acc
  }, [])
}

export function useAllActivities(account: string) {
  const { formatNumberOrString } = useLocalizationContext()
  const { activities, loading } = useAssetActivity()

  const localMap = useLocalActivities(account)
  const remoteMap = useMemo(() => {
    return parseRemoteActivities(activities, account, formatNumberOrString)
  }, [account, activities, formatNumberOrString])
  const updateCancelledTx = useTransactionCanceller()

  /* Updates locally stored pendings tx's when remote data contains a conflicting cancellation tx */
  useEffect(() => {
    if (!remoteMap) {
      return
    }

    Object.values(localMap).forEach((localActivity) => {
      if (!localActivity) {
        return
      }

      const cancelHash = findCancelTx({ localActivity, remoteMap, account })

      if (cancelHash) {
        updateCancelledTx({ id: localActivity.id, chainId: localActivity.chainId, cancelHash })
      }
    })
  }, [account, localMap, remoteMap, updateCancelledTx])

  const combinedActivities = useMemo(() => combineActivities(localMap, remoteMap ?? {}), [localMap, remoteMap])

  return { loading, activities: combinedActivities }
}

export function useOpenLimitOrders(account: string) {
  const { activities, loading } = useAllActivities(account)
  const openLimitOrders = activities.filter(
    (activity) =>
      activity.offchainOrderDetails?.routing === TradingApi.Routing.DUTCH_LIMIT &&
      activity.status === TransactionStatus.Pending,
  )

  return {
    openLimitOrders,
    loading,
  }
}

export function usePendingActivity() {
  const allPendingTransactions = usePendingTransactions()
  const pendingOrders = usePendingUniswapXOrders()

  // Filter out UniswapX orders from pendingTransactions to avoid double-counting
  // UniswapX orders are handled separately via pendingOrders
  const pendingTransactions = allPendingTransactions.filter((tx) => !isUniswapX(tx))
  // Pending limit orders shown in the limit sidebar
  const pendingOrdersWithoutLimits = pendingOrders.filter((order) => order.routing !== TradingApi.Routing.DUTCH_LIMIT)

  const hasPendingActivity = pendingTransactions.length > 0 || pendingOrdersWithoutLimits.length > 0
  const pendingActivityCount = pendingTransactions.length + pendingOrdersWithoutLimits.length
  const isOnlyUnichainPendingActivity =
    hasPendingActivity &&
    [...pendingTransactions, ...pendingOrdersWithoutLimits].every((tx) =>
      [UniverseChainId.Unichain, UniverseChainId.UnichainSepolia].includes(tx.chainId),
    )

  return { hasPendingActivity, pendingActivityCount, isOnlyUnichainPendingActivity }
}

export function useCancelOrdersGasEstimate(orders?: UniswapXOrderDetails[]): GasFeeResult {
  // Single order cancellation uses the same gas estimation as multi-order
  // The multi-order path handles both cases appropriately
  const cancelTransactionParams = useMemo(
    () =>
      orders && orders.length >= 1
        ? {
            orders: orders.filter(hasEncodedOrder).map((order) => ({
              encodedOrder: order.encodedOrder,
              routing: order.routing,
            })),
            chainId: orders[0].chainId,
          }
        : undefined,
    [orders],
  )

  const cancelTransaction = useCreateCancelTransactionRequest(cancelTransactionParams) ?? undefined
  return useTransactionGasFee(cancelTransaction, GasSpeed.Fast)
}
