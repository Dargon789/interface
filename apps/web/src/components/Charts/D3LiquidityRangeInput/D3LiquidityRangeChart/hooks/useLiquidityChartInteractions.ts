import { CHART_BEHAVIOR } from 'components/Charts/D3LiquidityRangeInput/D3LiquidityRangeChart/constants'
import { useChartPriceState } from 'components/Charts/D3LiquidityRangeInput/D3LiquidityRangeChart/store/selectors/priceSelectors'
import { ChartState } from 'components/Charts/D3LiquidityRangeInput/D3LiquidityRangeChart/store/types'
import { boundPanY } from 'components/Charts/D3LiquidityRangeInput/D3LiquidityRangeChart/utils/boundPanY'
import { calculateDynamicZoomMin } from 'components/Charts/D3LiquidityRangeInput/D3LiquidityRangeChart/utils/chartUtils'
import { ChartEntry } from 'components/Charts/LiquidityRangeInput/types'
import { useEffect } from 'react'

export function useLiquidityChartInteractions({
  svgRef,
  liquidityData,
  zoomLevel,
  panY,
  setChartState,
  dynamicZoomMin,
}: {
  svgRef: React.RefObject<SVGSVGElement | null>
  liquidityData: ChartEntry[]
  zoomLevel?: number
  panY?: number
  setChartState: (state: Partial<ChartState>) => void
  dynamicZoomMin?: number
}) {
  const { isFullRange } = useChartPriceState()

  useEffect(() => {
    const svgElement = svgRef.current
    if (!svgElement) {
      return
    }

    if (zoomLevel === undefined || panY === undefined || isFullRange) {
      return
    }

    // Setup wheel event handler for tick-based scrolling and pinch zoom
    const handleWheel = (event: WheelEvent) => {
      event.preventDefault()
      event.stopPropagation()

      const viewportHeight = svgElement.clientHeight

      // Check if this is a zoom gesture (Ctrl+wheel)
      if (event.ctrlKey) {
        const rect = svgElement.getBoundingClientRect()
        const centerY = event.clientY - rect.top

        // Calculate zoom based on wheel delta
        const deltaScale = -event.deltaY * CHART_BEHAVIOR.PINCH_DELTA_SCALE
        const zoomFactor = Math.max(
          CHART_BEHAVIOR.PINCH_ZOOM_FACTOR_MIN,
          Math.min(CHART_BEHAVIOR.PINCH_ZOOM_FACTOR_MAX, 1 + deltaScale),
        )

        const minZoom = dynamicZoomMin ?? calculateDynamicZoomMin(liquidityData.length)
        const newZoomLevel = Math.max(minZoom, Math.min(CHART_BEHAVIOR.ZOOM_MAX, zoomLevel * zoomFactor))

        // Adjust panY to keep mouse position fixed during zoom
        const zoomRatio = newZoomLevel / zoomLevel
        const newPanY = centerY - (centerY - panY) * zoomRatio

        const constrainedPanY = boundPanY({ panY: newPanY, viewportHeight, liquidityData, zoomLevel: newZoomLevel })

        setChartState({ zoomLevel: newZoomLevel, panY: constrainedPanY })
      } else {
        const scrollAmount = -event.deltaY
        const newPanY = panY + scrollAmount

        const constrainedPanY = boundPanY({ panY: newPanY, viewportHeight, liquidityData, zoomLevel })

        setChartState({ panY: constrainedPanY })
      }
    }

    // Add event listeners
    svgElement.addEventListener('wheel', handleWheel, { passive: false })

    // eslint-disable-next-line consistent-return
    return () => {
      svgElement.removeEventListener('wheel', handleWheel)
    }
  }, [liquidityData, zoomLevel, panY, dynamicZoomMin, isFullRange, setChartState, svgRef])
}
