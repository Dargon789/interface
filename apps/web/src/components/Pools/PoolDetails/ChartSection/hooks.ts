import { GraphQLApi } from '@universe/api'
import { ChartType } from 'components/Charts/utils'
import { SingleHistogramData } from 'components/Charts/VolumeChart/renderer'
import { ChartQueryResult, checkDataQuality, withUTCTimestamp } from 'components/Tokens/TokenDetails/ChartSection/util'
import { PDPChartQueryVars } from 'hooks/usePoolPriceChartData'
import { useMemo } from 'react'

export function usePDPVolumeChartData({
  variables,
}: {
  variables: PDPChartQueryVars & { addressOrId: string }
}): ChartQueryResult<SingleHistogramData, ChartType.VOLUME> {
  const { data, loading } = GraphQLApi.usePoolVolumeHistoryQuery({
    variables,
    skip: !variables.addressOrId || variables.addressOrId === '',
  })

  return useMemo(() => {
    const { historicalVolume } = data?.v2Pair ?? data?.v3Pool ?? data?.v4Pool ?? {}
    const entries =
      historicalVolume?.filter((amt): amt is GraphQLApi.TimestampedAmount => amt !== undefined).map(withUTCTimestamp) ??
      []

    const dataQuality = checkDataQuality({ data: entries, chartType: ChartType.VOLUME, duration: variables.duration })

    return { chartType: ChartType.VOLUME, entries, loading, dataQuality }
  }, [data?.v2Pair, data?.v3Pool, data?.v4Pool, loading, variables.duration])
}
