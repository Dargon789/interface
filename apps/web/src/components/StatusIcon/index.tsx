import sockImg from 'assets/svg/socks.svg'
import { CONNECTOR_ICON_OVERRIDE_MAP } from 'components/Web3Provider/constants'
import { useActiveAddresses, useActiveWallet } from 'features/accounts/store/hooks'
import { useHasSocks } from 'hooks/useSocksBalance'
import styled from 'lib/styled-components'
import { flexColumnNoWrap } from 'theme/styles'
import { Flex } from 'ui/src/components/layout'
import { breakpoints } from 'ui/src/theme'
import { AccountIcon } from 'uniswap/src/features/accounts/AccountIcon'
import { Platform } from 'uniswap/src/features/platforms/types/Platform'
import { isEVMAddress } from 'utilities/src/addresses/evm/evm'

const IconWrapper = styled.div<{ size?: number }>`
  position: relative;
  ${flexColumnNoWrap};
  align-items: center;
  justify-content: center;
  @media only screen and (min-width: ${breakpoints.xl}px) {
    margin-right: 4px;
  }
  & > img,
  span {
    height: ${({ size }) => (size ? size + 'px' : '32px')};
    width: ${({ size }) => (size ? size + 'px' : '32px')};
  }
  ${({ theme }) => theme.deprecated_mediaWidth.deprecated_upToMedium`
    align-items: flex-end;
  `};
`

const MINI_ICON_SIZE = 16

const MiniIconContainer = styled.div<{ side: 'left' | 'right'; size?: number; isIndicator?: boolean }>`
  position: absolute;
  display: flex;
  justify-content: center;
  align-items: center;
  width: ${({ size }) => size ?? MINI_ICON_SIZE + 'px'};
  height: ${({ size }) => size ?? MINI_ICON_SIZE + 'px'};
  bottom: ${({ size, isIndicator }) => `-${isIndicator ? 0 : (size ?? MINI_ICON_SIZE) / 4}px`};
  ${({ side, size, isIndicator }) => `${side === 'left' ? 'left' : 'right'}: -${isIndicator ? 0 : (size ?? MINI_ICON_SIZE) / 4}px`};
  border-radius: 50%;
  outline: 2px solid ${({ theme }) => theme.surface1};
  outline-offset: -0.1px;
  background-color: ${({ theme }) => theme.surface1};
  overflow: hidden;
  @supports (overflow: clip) {
    overflow: clip;
  }
`

const MiniImg = styled.img`
  width: ${MINI_ICON_SIZE + 'px'};
  height: ${MINI_ICON_SIZE + 'px'};
`

function Socks() {
  return (
    <MiniIconContainer side="left">
      <MiniImg src={sockImg} />
    </MiniIconContainer>
  )
}

function MiniWalletIcon({ platform }: { platform: Platform }) {
  const wallet = useActiveWallet(platform)
  if (!wallet) {
    return null
  }

  // TODO(APPS-8471): this should use useConnectedWallet() which returns connected WalletConnectorMeta, which is post-icon-override-map transformation
  const icon = CONNECTOR_ICON_OVERRIDE_MAP[wallet.name] ?? wallet.icon

  return (
    <MiniIconContainer side="right" data-testid="MiniIcon">
      <MiniImg src={icon} alt={`${wallet.name} icon`} />
    </MiniIconContainer>
  )
}

function MiniConnectedIndicator() {
  return (
    <MiniIconContainer isIndicator side="right" size={10}>
      <Flex backgroundColor="$statusSuccess" borderRadius="$roundedFull" height={10} width={10} />
    </MiniIconContainer>
  )
}

export default function StatusIcon({
  size = 16,
  showMiniIcons = true,
  showConnectedIndicator,
  address,
}: {
  size?: number
  showMiniIcons?: boolean
  showConnectedIndicator?: boolean
  address?: string
}) {
  const activeAddresses = useActiveAddresses()
  const hasSocks = useHasSocks()

  const addressToDisplay = address ?? activeAddresses.evmAddress ?? activeAddresses.svmAddress
  const platform = isEVMAddress(addressToDisplay) ? Platform.EVM : Platform.SVM

  return (
    <IconWrapper size={size} data-testid="StatusIconRoot">
      <AccountIcon address={addressToDisplay} size={size} />
      {showConnectedIndicator ? <MiniConnectedIndicator /> : showMiniIcons && <MiniWalletIcon platform={platform} />}
      {hasSocks && showMiniIcons && <Socks />}
    </IconWrapper>
  )
}
