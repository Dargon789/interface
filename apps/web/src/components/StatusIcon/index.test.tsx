import StatusIcon from 'components/StatusIcon'
import { useActiveAddresses, useActiveWallet } from 'features/accounts/store/hooks'
import { ExternalWallet } from 'features/accounts/store/types'
import { mocked } from 'test-utils/mocked'
import { render } from 'test-utils/render'

const ACCOUNT = '0x0'

vi.mock('uniswap/src/features/unitags/hooks', () => ({
  useUnitagByAddress: () => ({ unitag: undefined, loading: false }),
}))

vi.mock('../../hooks/useSocksBalance', () => ({
  useHasSocks: () => true,
}))

vi.mock('features/accounts/store/hooks')

vi.mock('uniswap/src/features/gating/hooks', () => ({
  useFeatureFlag: () => false,
  getFeatureFlag: () => false,
}))

describe('StatusIcon', () => {
  describe('with no account', () => {
    it('renders children in correct order', () => {
      mocked(useActiveWallet).mockReturnValue(undefined)
      mocked(useActiveAddresses).mockReturnValue({
        evmAddress: undefined,
        svmAddress: undefined,
      })
      const component = render(<StatusIcon />)
      expect(component.getByTestId('StatusIconRoot')).toMatchSnapshot()
      expect(component.queryByTestId('MiniIcon')).not.toBeInTheDocument()
    })
  })

  describe('with account', () => {
    beforeEach(() => {
      mocked(useActiveWallet).mockReturnValue({
        id: 'io.metamask',
        name: 'MetaMask',
        icon: '/src/assets/wallets/metamask-icon.svg',
      } as ExternalWallet)

      mocked(useActiveAddresses).mockReturnValue({
        evmAddress: ACCOUNT,
        svmAddress: undefined,
      })
    })

    it('renders children in correct order', () => {
      vi.spyOn(console, 'error').mockImplementation(() => null)

      const component = render(<StatusIcon />)
      expect(component.getByTestId('StatusIconRoot')).toMatchSnapshot()
      expect(component.getByTestId('MiniIcon')).toBeInTheDocument()
    })
  })

  it('renders without mini icons', async () => {
    const component = render(<StatusIcon showMiniIcons={false} />)
    expect(component.queryByTestId('MiniIcon')).not.toBeInTheDocument()
  })
})
