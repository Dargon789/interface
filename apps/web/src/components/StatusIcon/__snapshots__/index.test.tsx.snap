// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StatusIcon > with account > renders children in correct order 1`] = `
.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: column nowrap;
  -ms-flex-flow: column nowrap;
  flex-flow: column nowrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c0 > img,
.c0 span {
  height: 16px;
  width: 16px;
}

.c1 {
  position: absolute;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 16px;
  height: 16px;
  bottom: -4px;
  right: -4px;
  border-radius: 50%;
  outline: 2px solid #FFFFFF;
  outline-offset: -0.1px;
  background-color: #FFFFFF;
  overflow: hidden;
}

.c3 {
  position: absolute;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 16px;
  height: 16px;
  bottom: -4px;
  left: -4px;
  border-radius: 50%;
  outline: 2px solid #FFFFFF;
  outline-offset: -0.1px;
  background-color: #FFFFFF;
  overflow: hidden;
}

.c2 {
  width: 16px;
  height: 16px;
}

@media only screen and (min-width:1024px) {
  .c0 {
    margin-right: 4px;
  }
}

@media (max-width:960px) {
  .c0 {
    -webkit-align-items: flex-end;
    -webkit-box-align: flex-end;
    -ms-flex-align: flex-end;
    align-items: flex-end;
  }
}

@supports (overflow:clip) {
  .c1 {
    overflow: clip;
  }
}

@supports (overflow:clip) {
  .c3 {
    overflow: clip;
  }
}

<div
  class="c0"
  data-testid="StatusIconRoot"
  size="16"
>
  <div
    class="_display-flex _alignItems-stretch _flexBasis-auto _boxSizing-border-box _minHeight-0px _minWidth-0px _flexShrink-0 _flexDirection-column _backgroundColor-transparent _borderTopColor-transparent _borderRightColor-transparent _borderBottomColor-transparent _borderLeftColor-transparent _borderTopLeftRadius-t-radius-ro1041013639 _borderTopRightRadius-t-radius-ro1041013639 _borderBottomRightRadius-t-radius-ro1041013639 _borderBottomLeftRadius-t-radius-ro1041013639 _borderTopWidth-t-space-non101 _borderRightWidth-t-space-non101 _borderBottomWidth-t-space-non101 _borderLeftWidth-t-space-non101 _position-relative _width-16px _height-16px _borderBottomStyle-solid _borderTopStyle-solid _borderLeftStyle-solid _borderRightStyle-solid"
    data-testid="account-icon"
  >
    <div
      class="_opacity-0enter-0 _display-flex _alignItems-stretch _flexBasis-auto _boxSizing-border-box _position-relative _minHeight-0px _minWidth-0px _flexShrink-0 _flexDirection-column _opacity-1"
    >
      <span
        data-testid="unicon"
      >
        ðŸ”µ
      </span>
    </div>
  </div>
  <div
    class="c1"
    data-testid="MiniIcon"
  >
    <img
      alt="MetaMask icon"
      class="c2"
      src="/src/assets/wallets/metamask-icon.svg"
    />
  </div>
  <div
    class="c3"
  >
    <img
      class="c2"
      src="/src/assets/svg/socks.svg"
    />
  </div>
</div>
`;

exports[`StatusIcon > with no account > renders children in correct order 1`] = `
.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: column nowrap;
  -ms-flex-flow: column nowrap;
  flex-flow: column nowrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c0 > img,
.c0 span {
  height: 16px;
  width: 16px;
}

.c1 {
  position: absolute;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 16px;
  height: 16px;
  bottom: -4px;
  left: -4px;
  border-radius: 50%;
  outline: 2px solid #FFFFFF;
  outline-offset: -0.1px;
  background-color: #FFFFFF;
  overflow: hidden;
}

.c2 {
  width: 16px;
  height: 16px;
}

@media only screen and (min-width:1024px) {
  .c0 {
    margin-right: 4px;
  }
}

@media (max-width:960px) {
  .c0 {
    -webkit-align-items: flex-end;
    -webkit-box-align: flex-end;
    -ms-flex-align: flex-end;
    align-items: flex-end;
  }
}

@supports (overflow:clip) {
  .c1 {
    overflow: clip;
  }
}

<div
  class="c0"
  data-testid="StatusIconRoot"
  size="16"
>
  <div
    class="c1"
  >
    <img
      class="c2"
      src="/src/assets/svg/socks.svg"
    />
  </div>
</div>
`;
